File: backend/app/schemas/analytics.py
"""schemas/analytics.py — Phase 6 read-only analytics DTOs.

Shaped, denormalised metric bundles for the Overview + Analytics dashboards.
Everything here is computed deterministically by ``analytics_service`` (SQL /
maths) — that service is the single source of truth the dashboards read (and,
later, the AI query layer). Fields are additive by design: new metrics append,
they never restructure an existing bundle, so forms/columns can keep growing.
"""

from datetime import date
from typing import Literal

from pydantic import BaseModel, Field


class OverviewKpis(BaseModel):
    """Top-of-dashboard KPI bundle for one project (whole-project, unfiltered)."""

    # Pours
    pour_count: int = 0
    pour_volume_cum: float = 0.0
    # Quality (cube strength tests)
    test_count: int = 0
    pass_count: int = 0
    fail_count: int = 0
    critical_count: int = 0
    pass_rate_pct: float | None = None
    avg_strength_mpa: float | None = None
    # NCRs
    ncr_open: int = 0
    ncr_under_review: int = 0
    ncr_closed: int = 0
    avg_days_to_close: float | None = None
    # Dispatch / gate (trucks)
    truck_total: int = 0
    truck_accepted: int = 0
    truck_rejected: int = 0
    acceptance_pct: float | None = None


class GradeTrendPoint(BaseModel):
    period: str  # 'YYYY-MM'
    grade_name: str
    test_count: int
    pass_count: int
    pass_rate_pct: float | None = None


class StrengthBucket(BaseModel):
    label: str  # e.g. '<35', '35-40'
    count: int


class ResultBreakdown(BaseModel):
    status: str  # PASS / FAIL / CRITICAL_FAILURE / PENDING
    count: int


class QualityAnalytics(BaseModel):
    """Quality charts for the Analytics page, honouring the dimension filters."""

    grade_trend: list[GradeTrendPoint] = []
    strength_distribution: list[StrengthBucket] = []
    result_breakdown: list[ResultBreakdown] = []


class SupplierScore(BaseModel):
    supplier_id: int
    supplier_name: str
    pour_count: int = 0
    pour_volume_cum: float = 0.0
    test_count: int = 0
    pass_count: int = 0
    pass_rate_pct: float | None = None
    avg_strength_mpa: float | None = None


# ── Phase 5B: the four IS-456/10262 statistical charts (all filter-driven) ────


class RunPoint(BaseModel):
    """One individual 28-day result on the quality-control run chart."""

    test_date: str  # ISO date
    observed_mpa: float
    grade_name: str | None = None
    tower_name: str | None = None
    reference: str | None = None  # pour reference (kept for back-compat)
    sample_reference: str | None = None  # the cube number — the point's identity
    sample_id: int | None = None  # for deep-linking to traceability


class RunChart(BaseModel):
    """Chronological individual results + IS-456 control lines. Control lines are
    populated only when the data is a single grade (one fck)."""

    points: list[RunPoint] = []
    grade_name: str | None = None
    fck: float | None = None
    individual_min: float | None = None  # fck − 3
    target_mean: float | None = None  # fck + 1.65·σ
    mean: float | None = None


class CusumPoint(BaseModel):
    index: int  # cube number — chronological sequence, 1-based
    sample_reference: str | None = None
    sample_id: int | None = None
    test_date: str  # ISO date
    observed_mpa: float
    deviation: float  # observed − target mean
    cusum: float  # running cumulative sum of deviations


class CusumChart(BaseModel):
    """CUSUM control chart (IS-456 / SPC): the running sum of (observed − target
    mean) plotted by cube number. A sustained downward slope signals a fall in
    mean strength earlier than raw pass/fail. Single-grade (one target mean)."""

    points: list[CusumPoint] = []
    grade_name: str | None = None
    target_mean: float | None = None


class CurvePoint(BaseModel):
    x: float
    y: float


class DistributionCurve(BaseModel):
    """Normal distribution of the filtered strength dataset (IS-10262 basis)."""

    sample_count: int = 0
    mean: float | None = None
    std_dev: float | None = None
    fck: float | None = None
    curve: list[CurvePoint] = []  # sampled bell curve
    histogram: list[StrengthBucket] = []


class TargetMeanRow(BaseModel):
    grade_name: str
    fck: float
    target_mean: float  # fck + 1.65·σ
    actual_mean: float | None = None
    sample_count: int = 0


class TargetMeanChart(BaseModel):
    rows: list[TargetMeanRow] = []


class AgePoint(BaseModel):
    test_age_days: int
    observed_mpa: float
    required_mpa: float | None = None


class StrengthAgeChart(BaseModel):
    """Compressive strength vs age for a specific pour/element."""

    points: list[AgePoint] = []
    grade_name: str | None = None
    reference: str | None = None


class SupplierNcrCount(BaseModel):
    """NCRs raised against each supplier's pours, split by lifecycle + severity.

    ``open_count`` + ``closed_count`` = ``total`` (status is mutually exclusive);
    ``critical_count`` overlaps them — it counts the NCRs whose triggering cube
    test was a CRITICAL_FAILURE, regardless of whether they're open or closed.
    """

    supplier_id: int
    supplier_name: str
    open_count: int = 0
    closed_count: int = 0
    critical_count: int = 0
    total: int = 0


# ── Statistical tests (Student's t) ──────────────────────────────────────────
# Inference over cube-strength results: does a selection meet spec (one-sample),
# and do two selections differ (two-sample)? The confidence level defaults to
# 95% and is caller-adjustable. See app/core/statistics.py for the pure maths.

Alternative = Literal["two_sided", "greater", "less"]


class OneSampleTTest(BaseModel):
    """Result of testing a selection's mean cube strength against a reference."""

    sample_count: int
    mean: float
    std_dev: float
    std_error: float
    mu0: float
    mu0_basis: str  # 'fck' | 'target' | 'custom'
    grade_name: str | None = None
    values: list[float] = []  # the individual observed strengths (for plotting)
    t_statistic: float
    df: float
    p_value: float
    alternative: Alternative
    confidence: float
    ci_low: float
    ci_high: float
    significant: bool
    verdict: str


class GroupFilter(BaseModel):
    """The dimension selection defining one comparison group (two-sample test)."""

    grade_id: int | None = None
    tower_id: int | None = None
    supplier_id: int | None = None
    contractor_id: int | None = None
    date_from: date | None = None
    date_to: date | None = None
    label: str | None = None  # optional display name for the group


class TwoSampleRequest(BaseModel):
    group_a: GroupFilter
    group_b: GroupFilter
    confidence: float = Field(default=0.95, gt=0.5, lt=1.0)
    alternative: Alternative = "two_sided"


class GroupSummary(BaseModel):
    label: str
    sample_count: int
    mean: float | None = None
    std_dev: float | None = None
    values: list[float] = []  # the group's individual strengths (for plotting)


class TwoSampleTTest(BaseModel):
    """Welch two-sample t-test comparing two selections' mean cube strengths."""

    group_a: GroupSummary
    group_b: GroupSummary
    mean_diff: float
    t_statistic: float
    df: float
    p_value: float
    alternative: Alternative
    confidence: float
    ci_low: float
    ci_high: float
    significant: bool
    verdict: str

File: frontend/react-app/src/types/analytics.ts
// Phase 6: analytics.
// Read-only metric bundles from /projects/{id}/analytics/*. All fields are
// additive — new metrics append, so dashboards/forms can grow without churn.

export interface OverviewKpis {
  pour_count: number;
  pour_volume_cum: number;
  test_count: number;
  pass_count: number;
  fail_count: number;
  critical_count: number;
  pass_rate_pct: number | null;
  avg_strength_mpa: number | null;
  ncr_open: number;
  ncr_under_review: number;
  ncr_closed: number;
  avg_days_to_close: number | null;
  truck_total: number;
  truck_accepted: number;
  truck_rejected: number;
  acceptance_pct: number | null;
}

export interface GradeTrendPoint {
  period: string; // 'YYYY-MM'
  grade_name: string;
  test_count: number;
  pass_count: number;
  pass_rate_pct: number | null;
}

export interface StrengthBucket {
  label: string;
  count: number;
}

export interface ResultBreakdown {
  status: string;
  count: number;
}

export interface QualityAnalytics {
  grade_trend: GradeTrendPoint[];
  strength_distribution: StrengthBucket[];
  result_breakdown: ResultBreakdown[];
}

export interface SupplierScore {
  supplier_id: number;
  supplier_name: string;
  pour_count: number;
  pour_volume_cum: number;
  test_count: number;
  pass_count: number;
  pass_rate_pct: number | null;
  avg_strength_mpa: number | null;
}

export interface SupplierNcrCount {
  supplier_id: number;
  supplier_name: string;
  open_count: number;
  closed_count: number;
  critical_count: number;
  total: number;
}

export interface QualityFilters {
  date_from?: string; // ISO date
  date_to?: string; // ISO date
  grade_id?: number;
  supplier_id?: number;
  tower_id?: number;
  contractor_id?: number;
  component_id?: number;
  sample_id?: number;
}

// ── Phase 5B: the four IS-456/10262 statistical charts ──────────────────────

export interface RunPoint {
  test_date: string;
  observed_mpa: number;
  grade_name: string | null;
  tower_name: string | null;
  reference: string | null;
  sample_reference: string | null; // the cube number — the point's identity
  sample_id: number | null;
}

export interface CusumPoint {
  index: number; // cube number (1-based sequence)
  sample_reference: string | null;
  sample_id: number | null;
  test_date: string;
  observed_mpa: number;
  deviation: number; // observed − target mean
  cusum: number; // running cumulative sum
}
export interface CusumChart {
  points: CusumPoint[];
  grade_name: string | null;
  target_mean: number | null;
}
export interface RunChart {
  points: RunPoint[];
  grade_name: string | null;
  fck: number | null;
  individual_min: number | null;
  target_mean: number | null;
  mean: number | null;
}

export interface CurvePoint { x: number; y: number }
export interface DistributionCurve {
  sample_count: number;
  mean: number | null;
  std_dev: number | null;
  fck: number | null;
  target_mean: number | null; // RMC design target, else IS-10262 fck+1.65σ
  curve: CurvePoint[];
  histogram: StrengthBucket[];
}

// ── Graphical summary (Minitab-style descriptive report) ────────────────────
export interface HistogramBar {
  bin_low: number;
  bin_high: number;
  count: number;
}
export interface ProbPoint {
  value: number; // ordered observed strength
  theoretical: number; // theoretical normal quantile
}
export interface GraphicalSummary {
  sample_count: number;
  grade_name: string | null;
  fck: number | null;
  mean: number | null;
  std_dev: number | null;
  variance: number | null;
  skewness: number | null;
  kurtosis: number | null;
  minimum: number | null;
  q1: number | null;
  median: number | null;
  q3: number | null;
  maximum: number | null;
  ci_confidence: number;
  ci_mean_low: number | null;
  ci_mean_high: number | null;
  ad_statistic: number | null;
  ad_p_value: number | null;
  is_normal: boolean | null;
  bin_width: number | null;
  histogram: HistogramBar[];
  fit_curve: CurvePoint[]; // fitted normal PDF (density)
  kde_curve: CurvePoint[]; // Gaussian KDE (density)
  prob_points: ProbPoint[];
}

// ── Outlier scan (modified Thompson τ) ──────────────────────────────────────
export interface OutlierPoint {
  index: number; // 1-based chronological position
  value: number;
  is_outlier: boolean;
}
export interface OutlierAnalysis {
  sample_count: number;
  grade_name: string | null;
  mean: number | null;
  std_dev: number | null;
  outlier_count: number;
  clean_mean: number | null;
  clean_std_dev: number | null;
  tau: number | null;
  threshold: number | null; // τ·S — rejection distance from the mean
  points: OutlierPoint[];
  outliers: number[];
}

export interface TargetMeanRow {
  grade_name: string;
  fck: number;
  target_mean: number;
  actual_mean: number | null;
  sample_count: number;
}
export interface TargetMeanChart { rows: TargetMeanRow[] }

export interface AgePoint {
  test_age_days: number;
  observed_mpa: number;
  required_mpa: number | null;
}
export interface StrengthAgeChart {
  points: AgePoint[];
  grade_name: string | null;
  reference: string | null;
}

// ── Statistical tests (Student's t) ─────────────────────────────────────────

export type Alternative = 'two_sided' | 'greater' | 'less';
export type TTestBasis = 'fck' | 'target' | 'custom';

export interface OneSampleParams {
  grade_id?: number;
  tower_id?: number;
  supplier_id?: number;
  contractor_id?: number;
  basis?: TTestBasis;
  mu0?: number;
  confidence?: number;
  alternative?: Alternative;
  date_from?: string;
  date_to?: string;
}

export interface OneSampleTTest {
  sample_count: number;
  mean: number;
  std_dev: number;
  std_error: number;
  mu0: number;
  mu0_basis: TTestBasis;
  grade_name: string | null;
  values: number[];
  t_statistic: number;
  df: number;
  p_value: number;
  alternative: Alternative;
  confidence: number;
  ci_low: number;
  ci_high: number;
  significant: boolean;
  verdict: string;
}

export interface GroupFilter {
  grade_id?: number;
  tower_id?: number;
  supplier_id?: number;
  contractor_id?: number;
  date_from?: string;
  date_to?: string;
  label?: string;
}

export interface TwoSampleRequest {
  group_a: GroupFilter;
  group_b: GroupFilter;
  confidence: number;
  alternative: Alternative;
}

export interface GroupSummary {
  label: string;
  sample_count: number;
  mean: number | null;
  std_dev: number | null;
  values: number[];
}

export interface TwoSampleTTest {
  group_a: GroupSummary;
  group_b: GroupSummary;
  mean_diff: number;
  t_statistic: number;
  df: number;
  p_value: number;
  alternative: Alternative;
  confidence: number;
  ci_low: number;
  ci_high: number;
  significant: boolean;
  verdict: string;
}
